### DISCUSSION TITLE
event handlers registered using `$wire.on` are not fired

### DISCUSSION THREAD

**Discussion by jhcarl0814** [livewire/livewire#5407] [7. Bugs (archived)] [answered]
minimal  reproducible example:
resources\views\livewire\form-enquiry.blade.php
```blade
<div>
    <form x-data="{ }" @submit.prevent="
        console.log($wire);
        console.log(Livewire);
        $wire.on('foo', (e) => {
            console.log(1,e); // not executed
        });
        Livewire.on('foo', (e) => {
            console.log(2,e);
        });
        $wire.emit('foo', new Date().toTimeString());
        ">
        <button type="submit">{{$repo->send_button_text }}</button>
    </form>
</div>
```

**Comment by joshhanley** [comment]
@jhcarl0814 that's correct, you should register your listeners using `Livewire.on`, see here for details https://laravel-livewire.com/docs/2.x/events#in-js

Did you see `$wire.on` somewhere in the docs? If so we will have to update that.

Hope this helps!

**Reply by jhcarl0814** [reply]
Here is the `$wire` documentation link: [Interacting With Livewire From Alpine: $wire](https://laravel-livewire.com/docs/2.x/alpine-js#interacting-with-livewire-from-alpine)
```js
// Listening for a Livewire event emitted called "some-event"
$wire.on('some-event', (foo, bar) => {})
```

Although the docs does not say, it seems:

- `Livewire.on` in `<script>`s is registering globally so the callback persists even if the livewire component containing `<script>`s is removed;
- `$wire.on` in `alpine attributes' value`s is scoped to the livewire component so the callback will be deregistered when livewire component is removed;

Unfortunately, it is not possible to test if my imagination is correct because `$wire.on` is not working. But it would be great if you could double check their original intended use cases or provide a hint on how to fix this (which files or which functions should I look at).

**Reply by joshhanley** [reply]
@jhcarl0814 ah yep, thanks, I missed that! So I just did a quick test and it works happily for me:

```blade
<div x-data x-init="$wire.on('test', () => console.log('test'))">
    <button type="button" wire:click="$emit('test')">Test</button>
</div>
```

Can you test that and see if it works for you?

**Reply by jhcarl0814** [reply]
@joshhanley Thanks for the example! It works for me. But it seems only `wire:browser_event_name="$emit('livewire_event_name', ...)"` can trigger `$wire.on('livewire_event_name', fn)`, which is unfortunate because in `wire:browser_event_name` you can not write arbitrary js code but can only do 1 action. Switching to `x-init`/`x-on:browser_event_name` allows you to write arbitrary js code but it can not trigger `$wire.on('livewire_event_name', fn)`.

```blade
<div>
    <form class="flex flex-col" x-data="{ }"
        x-init="
            console.log(0);
            $wire.on('test', (e) => console.log(1,e));
            Livewire.on('test', (e) => console.log(2,e));
            $wire.emit('test', 3);
            Livewire.emit('test', 4);"
        @submit.prevent="
            console.log(5);
            $wire.on('test', (e) => console.log(6,e));
            Livewire.on('test', (e) => console.log(7,e));
            $wire.emit('test', 8);
            Livewire.emit('test', 9);">

<pre>
0
2 3
2 4
</pre>

<button type="submit">submit</button>
<pre>
5
2 8
7 8 *n
2 9
7 9 *n
</pre>

<button type="button" wire:click="$emit('test', 10)">wire:click $emit</button>
<pre>
1 10
6 10 *n
2 10
7 10 *n
</pre>

<button type="button" @click="$wire.emit('test', 11);">@click $wire.emit</button>
<pre>
2 11
7 11 *n
</pre>

<button type="button" @click="Livewire.emit('test', 12);">@click Livewire.emit</button>
<pre>
2 12
7 12 *n
</pre>

    </form>
</div>
```

|   | `$wire.on('livewire_event_name', fn)` | `Livewire.on('livewire_event_name', fn)` |
| --- | --- | --- |
| `wire:browser_event_name="$emit('livewire_event_name', ...)"` | yes | yes |
| `x-init`/`x-on:browser_event_name="... $wire.emit('livewire_event_name', ...); ..."` | no | yes |
| `x-init`/`x-on:browser_event_name="... Livewire.emit('livewire_event_name', ...); ..."` | no | yes |

**Comment by joshhanley** [comment, marked as answer]
@jhcarl0814 ah yeah, I see where the issue is. Yeah that is a bug, when using `$wire.emit...()` it's not firing any of the scoped listeners, only the global ones.

To fix, this `component.scopedListeners.call(...args)` would need to be added to the top of this closure here
https://github.com/livewire/livewire/blob/0edabd715f0ea241a883694669185f41c54567df/js/component/index.js#L657-L662

Which is similar to how scoped listeners get called here

https://github.com/livewire/livewire/blob/0edabd715f0ea241a883694669185f41c54567df/js/component/index.js#L310-L322

**Reply by jhcarl0814** [reply]
@joshhanley Thank you for the solution!
After adding `component.scopedListeners.call(...args)`, `$wire.emit` fires scoped listeners.
![image](https://user-images.githubusercontent.com/5171262/206319504-a2b6b48a-0793-46dd-95bf-1278f660960f.png)


But `Livewire.emit` still does not fire scoped listeners.

```blade
<div>
    <form class="flex flex-col" x-data="{ }"
        x-init="
            console.log(0);
            $wire.on('test', (e) => console.log(1,e));
            Livewire.on('test', (e) => console.log(2,e));
            $wire.emit('test', 3);
            Livewire.emit('test', 4);"
        @submit.prevent="
            console.log(5);
            $wire.on('test', (e) => console.log(6,e));
            Livewire.on('test', (e) => console.log(7,e));
            $wire.emit('test', 8);
            Livewire.emit('test', 9);">

<pre>
0
1 3
2 3
2 4
</pre>

<button type="submit">submit</button>
<pre>
5
1 8
6 8 *n
2 8
7 8 *n
2 9
7 9 *n
</pre>

<button type="button" wire:click="$emit('test', 10)">wire:click $emit</button>
<pre>
1 10
6 10 *n
2 10
7 10 *n
</pre>

<button type="button" @click="$wire.emit('test', 11);">@click $wire.emit</button>
<pre>
1 11
6 11 *n
2 11
7 11 *n
</pre>

<button type="button" @click="Livewire.emit('test', 12);">@click Livewire.emit</button>
<pre>
2 12
7 12 *n
</pre>

    </form>
</div>
```

|   | `$wire.on('livewire_event_name', fn)` | `Livewire.on('livewire_event_name', fn)` |
| --- | --- | --- |
| `wire:browser_event_name="$emit('livewire_event_name', ...)"` | yes | yes |
| `x-init`/`x-on:browser_event_name="... $wire.emit('livewire_event_name', ...); ..."` | yes (add `component.scopedListeners.call(...args);` in closure after `/^emit.*/`) | yes |
| `x-init`/`x-on:browser_event_name="... Livewire.emit('livewire_event_name', ...); ..."` | no | yes |

**Reply by joshhanley** [reply]
@jhcarl0814 yeah that's correct, scoped listeners will only fire from methods called within the component itself (ie using `$emit`).

Where as using `Livewire.emit` is a global emit, so it's not scoped, hence why it doesn't fire. 🙂

**Reply by jhcarl0814** [reply]
@joshhanley Thank you for the clarification! It would be better if those concepts are formalized/reflected in the docs.

**Reply by joshhanley** [reply]
@jhcarl0814 no worries, and yep, agreed! PR's to the docs are welcome 🙂 https://github.com/livewire/docs
